### PR TITLE
[WIP] Import AST from JSON format (try 2)

### LINKED ISSUE #2096: AST: import from json
It should be possible to import the AST from the new json format (this is also a good way to test that we did not forget anything). The idea is that after parsing and type checking, all information in the compiler can be exported and then later re-imported into a new instance of the compiler and produce the same bytecode.
This opens the door for things like mutation testing.

### PR DESCRIPTION
**Pull request by djudjuu**
I rebased the branch, included the CR's from the original branch and cleaned up the code.

Replaces #2328.

Closes #2096.

### CODE REVIEW

**Comment by djudjuu** [comment]
There is one bit about which i am unhappy and would like your help:
When the code compilation leads to warnings, the method SourceReferenceFormatter::printExceptionInformation() wants to output the sourceLocation, which will always print the entire Json-file. This is not really helpful.
Also, as InlineAssemblyeCode needs its own scanner, I needed to add an extra variable m_currentSourceName to ASTJsonImporter which is only used once. 

I think the most appropriate way would be deal with the the generated warninsgs would be to output only the type of warning without the entire code. I am unsure about how to (elegantly) adjust the printExceptionInformation to that end.
i guess it's easiest to discuss this on a call...

**Comment by axic** [comment]
Is this replacing #2328?

**Comment by djudjuu** [comment]
Yes, I don't have write access anymore so I had to do it that way..
Also couldn't change the label to needsReview

File: Changelog.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ Features:
  * Assembly: renamed ``SHA3`` to ``KECCAK256``.
  * Assembly: Display auxiliary data in the assembly output.
  * Assembly: Add ``CREATE2`` (EIP86), ``STATICCALL`` (EIP214), ``RETURNDATASIZE`` and ``RETURNDATACOPY`` (EIP211) instructions.
- * AST: export all attributes to JSON format.
+ * AST: import/export the AST from/to Json format
```
Comment (axic) [inline Changelog.md]: Please add a new changelog entry since this release is done.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 2]
```diff
@@ -787,8 +787,10 @@ string ASTJsonConverter::functionCallKind(FunctionCallKind _kind)
 		return "typeConversion";
 	case FunctionCallKind::StructConstructorCall:
 		return "structConstructorCall";
+	case FunctionCallKind::Unset:
+		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("FunctionCallKind was not set."));
 	default:
-		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Unknown kind of function call ."));
+		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Unknown kind of function call."));
```
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp]: Do you want to create a separate PR for these two changes above? Could be merged right away.

File: libsolidity/interface/CompilerStack.cpp [hunk 3]
```diff
@@ -775,13 +807,13 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
...
 	meta["version"] = 1;
-	meta["language"] = "Solidity";
+	meta["language"] = m_importedSources ? "SolidityAST" : "Solidity";
 	meta["compiler"]["version"] = VersionStringStrict;
 
 	meta["sources"] = Json::objectValue;
 	for (auto const& s: m_sources)
 	{
-		solAssert(s.second.scanner, "Scanner not available");
+		solAssert(s.second.scanner,"");
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: Any point removing this?

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 4]
```diff
@@ -0,0 +1,844 @@
...
+#include <libsolidity/parsing/Token.h>
+#include <libsolidity/inlineasm/AsmParser.h>
+#include <libsolidity/interface/ErrorReporter.h>
+
+using namespace std;
+
+namespace
+{
+template<class T>
+dev::solidity::ASTPointer<T> castPointer(dev::solidity::ASTPointer<dev::solidity::ASTNode> _ast)
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.cpp]: If you move these below `using namespace solidity` then the prefixes are not needed.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 5]
```diff
@@ -0,0 +1,844 @@
...
+		start,
+		end,
+		m_sourceLocations[stoi(pos[2])]
+	);
+}
+
+//helper functions
+ASTPointer<ASTString> ASTJsonImporter::memberAsASTString(Json::Value const& _node, string const& _name)
+{
+	astAssert(member(_node, _name).isString(), "field " + _name + " must be of type string.");
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.cpp]: Perhaps:
```
Json::Value value = member(_node, _name);
astAssert(value.isSting(), "...");
return make_shared<ASTString>(value.asString());
```

and the same for `memberAsBool` makes it clearer.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 6]
```diff
@@ -0,0 +1,844 @@
...
+	astAssert(member(_node, _name).isBool(), "field " + _name + " must be of type boolean.");
+	return _node[_name].asBool();
+}
+
+Json::Value ASTJsonImporter::member(Json::Value const& _node, string const& _name)
+{
+	if (_node.isMember(_name))
+		return _node[_name];
+	else
+		BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("Node '" + _node["nodeType"].asString() + "' (id " + _node["id"].asString() + ") is missing field '" + _name + "'."));
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.cpp]: Please change this to
```
astAssert(_node.isMember(_name), "...");
return _node[_name];
```

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 7]
```diff
@@ -0,0 +1,844 @@
...
+	return _json[_name].isString() ? memberAsASTString(_json, _name) : nullptr;
+}
+
+Token::Value ASTJsonImporter::scanSingleToken(Json::Value _node)
+{
+	Scanner scanner(CharStream(_node.asString()));
+	if (scanner.peekNextToken() == Token::EOS)
+		return scanner.currentToken();
+	else
+		BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("Token string is too long."));
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.cpp]: Could be changed to `astAssert(peekNextToken ...)`.

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 8]
```diff
@@ -0,0 +1,844 @@
...
+}
+
+ASTPointer<ASTNode> ASTJsonImporter::createStructDefinition(Json::Value const& _node)
+{
+	std::vector<ASTPointer<VariableDeclaration>> members;
+	for (auto& member: _node["members"])
+		members.push_back(createVariableDeclaration(member));
+	return createASTNode<StructDefinition>(
+		_node,
+		make_shared<ASTString>(_node["name"].asString()),
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.cpp]: `memberAsASTString`?

**Comment by axic** [comment]
@djudjuu pushed this changeset to the original branch to see which comments were left open (there are a couple), can you please check them there?

**Comment by djudjuu** [comment]
Hi @axic ,
I took another go at it, included all changes that were still missing, rebased it onto the current develop (in the process I had to squash my many commits into fewer and do a force push, although it meant that some of your comments are gone now, but otherwise i'd gone nuts. (They were some whitespace errors and some refactoring, no real design choices or the like....). 
I also included the comments from #2328.

Also, I put a question in the comment in CompilerStack.cpp l. 846....it's in the metadata creation function, which is also the place where I felt the most unsure about my choices.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 9]
```diff
@@ -770,6 +770,8 @@ string ASTJsonConverter::functionCallKind(FunctionCallKind _kind)
 		return "typeConversion";
 	case FunctionCallKind::StructConstructorCall:
 		return "structConstructorCall";
+	case FunctionCallKind::Unset:
+		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("FunctionCallKind was not set."));
```
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp]: Should be a `solAssert`.

File: solc/CommandLineInterface.cpp [hunk 10]
```diff
@@ -781,8 +850,22 @@ bool CommandLineInterface::processInput()
...
+		if (m_args.count(g_argImportAst))
+		{
+			map<string, Json::Value const*> sourceJsons = parseAstFromInput();
+			//feed AST to compiler
+			m_compiler->reset(false);
+			//use the compiler's analyzer to annotate, typecheck, etc...
+			if (!m_compiler->importASTs(sourceJsons))
+				BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("Import of the AST failed"));
+			if (!m_compiler->analyze())
+				BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("Analysis of the AST failed"));
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: I guess these should use `astAssert`?

File: test/libsolidity/ASTJSON.cpp [hunk 11]
```diff
@@ -254,10 +256,41 @@ BOOST_AUTO_TEST_CASE(documentation)
...
+	assert(c.compile());
+	Json::Value newJson = ASTJsonConverter(false, c.sourceIndices()).toJson(c.ast("a"));
+	assert(newJson == originalJson);
+	string newBinary = dev::test::bytecodeSansMetadata(c.object(c.contractNames()[0]).toHex());
+	// Note: Bytecode comparison will only succeed if the contracts don't import/inherit other contracts, because
+	// the other contract's bytecode will contain it's metadata, whose 'source'-field is usually
+	// generated from the solidity-code, but after importing will be generated from the Json.
+	// Therefore, in order to test more complicated contracts, the bytecodeSansMetadata-function needs to be extended to
+	// also remove metadatabytecode not only from the end but also from the middle of a given string.
+	assert(newBinary == originalBinary);
```
Comment (axic) [inline test/libsolidity/ASTJSON.cpp]: `BOOST_CHECK_EQUAL` should be used instead of asserts here.

**Comment by axic** [comment]
Sorry for the delay. Thanks! Pushed your version to #2328 and can confirm all the questions were addressed.

File: test/libsolidity/ASTJSON.cpp (hunk 11 above)
Comment (axic) [reply test/libsolidity/ASTJSON.cpp]: Sorry I meant this for every `assert` in that function (there are a couple more above).

Comment (djudjuu) [reply test/libsolidity/ASTJSON.cpp]: oh, should have been obvious, thx. done now.

**Comment by chriseth** [comment]
*cheer*

**Comment by djudjuu** [comment]
I (finally) updated the missing pieces, yet I don't know what to do about the appveyor error. @chriseth can you help?

File: solc/CommandLineInterface.h [hunk 12]
```diff
@@ -73,6 +73,9 @@ class CommandLineInterface
 
 	/// Fills @a m_sourceCodes initially and @a m_redirects.
 	void readInputFilesAndConfigureRemappings();
+	/// Tries to read the ASTs from @ m_sourceCodes and to generate the input for
+	/// the compiler's importASTs()-function.
+	std::map<std::string, Json::Value const*> parseAstFromInput(); //use const here?
```
Comment (axic) [inline solc/CommandLineInterface.h:78]: Should use `const` if it isn't changing the state.

File: Changelog.md [hunk 13]
```diff
@@ -204,6 +204,7 @@ Bugfixes:
 Features:
  * Syntax Checker: Deprecated "throw" in favour of require(), assert() and revert().
  * Type Checker: Warn if a local storage reference variable does not explicitly use the keyword ``storage``.
+ * AST: import a Json to AST.
```
Comment (axic) [inline Changelog.md:207]: This is in the wrong version, also please capitalise to JSON.

File: libsolidity/ast/ASTJsonImporter.h [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+ * Converter of the AST from JSON sformat to ASTNode
+ */
+class ASTJsonImporter
+{
+public:
+	/// Create an importer to import a given abstract syntax tree in Json format to an ASTNode
+	/// @a _sourceList used to provide source names for the ASTs
+	ASTJsonImporter(std::map<std::string, Json::Value const*> _sourceList);
+
+	/// converts the
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.h:46]: Missing sentence?

File: libsolidity/ast/ASTJsonImporter.h [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+	std::map<std::string, ASTPointer<SourceUnit>> jsonToSourceUnit();
+
+private:
+	template <typename T, typename... Args>
+	ASTPointer<T> createASTNode(Json::Value const& _node, Args&&... _args);
+	SourceLocation const createSourceLocation(Json::Value const& _node);
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.h:52]: Can this really be const?

File: libsolidity/ast/ASTJsonImporter.h [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+	ASTPointer<ASTString> memberAsASTString(Json::Value const& _node, std::string const& _name);
+	bool memberAsBool(Json::Value const& _node, std::string const& _name);
+	Declaration::Visibility visibility(Json::Value const& _node);
+	StateMutability stateMutability(Json::Value const& _node);
+	VariableDeclaration::Location location(Json::Value const& _node);
+	ContractDefinition::ContractKind contractKind(Json::Value const& _node);
+	Literal::SubDenomination subdenomination(Json::Value const& _node);
+	Token::Value literalTokenKind(Json::Value const& _node);
+	Token::Value scanSingleToken(Json::Value _node);
+	std::map<std::string, Json::Value const*> m_sourceList;
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.h:115]: Can you separate the members from the helpers with a space?

File: libsolidity/ast/ASTJsonImporter.h [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+	ASTPointer<NewExpression> createNewExpression(Json::Value const& _node);
+	ASTPointer<MemberAccess> createMemberAccess(Json::Value const& _node);
+	ASTPointer<IndexAccess> createIndexAccess(Json::Value const& _node);
+	ASTPointer<Identifier> createIdentifier(Json::Value const& _node);
+	ASTPointer<ElementaryTypeNameExpression> createElementaryTypeNameExpression(Json::Value const& _node);
+	ASTPointer<ASTNode> createLiteral(Json::Value const& _node);
+
+	template<class T>
+	ASTPointer<T> nullOrCast(Json::Value _json);
+	ASTPointer<ASTString> nullOrASTString(Json::Value _json, std::string const& _name);
```
Comment (axic) [inline libsolidity/ast/ASTJsonImporter.h:104]: Can any of these be `const`? Can the `_json` parameter be a `const&`?

**Review by axic — CHANGES_REQUESTED** [review]
Please rebase also.

**Comment by ekpyron** [comment]
@djudjuu Are you planning on continuing with this? I'm just starting to look into it and I could help you in rebasing it (it has been dormant for quite a while now) and finishing it - or I could just take it over.

**Comment by djudjuu** [comment]
Hi @ekpyron ,

I have been meaning to continue working on this for ages. So yes, I am very interested to continue working on it, unfortunately my other work has been getting in the way.

If you'd be down to pair on it a little bit or just rebase it for a start, that'd be cool.

I am really keen on seeing it done  because I want to start
experimenting with a mutation-testing framework :)
So far my plan was to start working on it after december 5th, when I go
into vacation, but if you're down, we could start addressing it before that
too.

**Comment by ekpyron** [comment]
@djudjuu Ok, sounds great! I'll probably look into rebasing it sometime this week then and after that we can decide how to proceed.

**Comment by axic** [comment]
> Submodule deps added at b3db89

Should be removed.

**Comment by ekpyron** [comment]
I looked into rebasing this a while ago, but the PR seems to be in a somewhat messy state (e.g. there's merge conflict markers in some of the commits that are only removed later, etc.) - so it may make sense to recreate it from scratch instead.

**Comment by axic** [comment]
I tried rebasing it yesterday, but due to the merge commits it is rather hard. Tried rebasing it on an older head just to remove the merge and then squash, but haven't finished it yet.

**Comment by chriseth** [comment]
It is probably easier to just create this from scratch.

**Comment by djudjuu** [comment]
Hi @ekpyron,

I restarted work on this again over here: https://github.com/djudjuu/solidity/pull/1

let me know if you want to help out!

File: solc/CommandLineInterface.h (hunk 12 above)
Comment (djudjuu) [reply solc/CommandLineInterface.h:78]: hmm, it does set ```m_sourceCodes```

also it reads the input file into a newly created Json::Value which after that never changes, thats why i though that Json::Value have a *const**. 

not sure, what that implies...

File: libsolidity/ast/ASTJsonImporter.h (hunk 15 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonImporter.h:52]: i though so, as the json that is passed in is never modified.

File: libsolidity/ast/ASTJsonImporter.h (hunk 17 above)
Comment (djudjuu) [reply libsolidity/ast/ASTJsonImporter.h:104]: yes. done
